sw/device/lib/dif/dif_spi_device.c

```diff
@@ -566,87 +559,32 @@ static size_t spi_memcpy(const dif_spi_device_t *spi, fifo_ptrs_t *fifo,
     ptr = &fifo->write_ptr;
   }
 
-  uint16_t total_bytes_available = bytes_left;
-
-  // First, bring the fifo write pointer into word alignment, so we can do
-  // simple full-word I/O rather than partial word I/O; note that the FIFO
-  // SRAM is not guaranteed to support subword writes.
-  ptrdiff_t misalignment = misalignment32_of(ptr->offset);
-  if (misalignment != 0) {
-    // The number of bytes missing to bring `ptr->offset` back into alignment.
-    // For exmaple, 0x3 has misalignment of 3 and realignment of 1.
-    ptrdiff_t realignment = sizeof(uint32_t) - misalignment;
-    // Note that we might be doing less I/O than the misalignment requires; we
-    // might be off by a single byte, but not have the full three bytes for full
-    // realignment.
-    if (realignment > bytes_left) {
-      realignment = bytes_left;
-    }
-
-    // Converts `offset`, which points to a subword boundary, to point to the
-    // start of the current word it points into.
-    ptrdiff_t current_word_offset = ptr->offset - misalignment;
-    uint32_t current_word =
-        mmio_region_read32(spi->base_addr, SPI_DEVICE_BUFFER_REG_OFFSET +
-                                               fifo_base + current_word_offset);
-
-    // Act on only to a suffix of `current_word`, corresponding to the necessary
-    // realignment.
-    uint8_t *current_byte = ((uint8_t *)&current_word) + misalignment;
-    if (is_recv) {
-      memcpy(byte_buf, current_byte, realignment);
-    } else {
-      // When sending, we need to commit the staged write.
-      memcpy(current_byte, byte_buf, realignment);
-      mmio_region_write32(spi->base_addr, SPI_DEVICE_BUFFER_REG_OFFSET +
-                                              fifo_base + current_word_offset,
-                          current_word);
-    }
-
-    fifo_ptr_increment(ptr, realignment, fifo_len);
-    byte_buf += realignment;
-    bytes_left -= realignment;
-  }
-
-  // Now, we just do full word I/O until we run out of stuff to act on.
+  // `mmio_region_memcpy_*_mmio32` functions assume sequential memory access
+  // while the SPI device uses a circular buffer. Therefore, we split the copy
+  // operation into chunks that access the device buffer sequentially.
```

Comment:
Huh, is there something I missed in my original implementation?

Reply:
Apologies if I'm repeating myself since we talked briefly about this. You original implementation did not have any issues. This is just because the original implementation was obviously aware of the SPI buffer pointers and they were updated inside the loop. Since the new function assumes sequential access, we may have to call it one more time if the pointers wrap around.